File: NeetCode/Trie/Extra_Characters_In_String/extra__characters__in__string.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        pft = PrefixTree()
        for word in dictionary:
            pft.insert(word)

        n = len(s)

        dp = [0 for _ in range(n + 1)]

        for index in range(n - 1, -1, -1):
            dp[index] = 1 + dp[index + 1]

            ptr = pft.root

            for j in range(index, n):
                ch = s[j]

                if ch not in ptr.children:
                    break
                ptr = ptr.children[ch]

                if ptr.terminal:
                    dp[index] = min(dp[index], dp[j + 1])

        return dp[0]


class Node:
    def __init__(self):
        self.children = {}
        self.terminal = False


class PrefixTree:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        ptr = self.root
        for w in word:
            if w not in ptr.children:
                ptr.children[w] = Node()
            ptr = ptr.children[w]
        ptr.terminal = True


if __name__ == "__main__":
    sol = Solution()

    print("Running Solution...")

File: NeetCode/Trie/Extra_Characters_In_String/Extra_Characters_In_String.cpp
// Extra_Characters_In_String.cpp
//
// Compile locally with:
// clang++ Extra_Characters_In_String.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
private:
    struct TrieNode
    {
        bool terminal = false;
        unordered_map<char, TrieNode *> children;
    };

    TrieNode *root = new TrieNode();

    void insert(const string &s)
    {
        TrieNode *ptr = root;
        for (char ch : s)
        {
            if (ptr->children.count(ch) == 0)
            {
                ptr->children[ch] = new TrieNode();
            }
            ptr = ptr->children[ch];
        }
        ptr->terminal = true;
    }

public:
    int minExtraChar(string s, vector<string> &dictionary)
    {
        for (auto &word : dictionary)
        {
            insert(word);
        }

        int n = s.size();
        vector<int> dp(n + 1, 0);
        for (int index = n - 1; index >= 0; index--)
        {
            dp[index] = 1 + dp[index + 1];
            TrieNode *ptr = root;
            for (int j = index; j < n; j++)
            {
                if (ptr->children.count(s[j]) == 0)
                {
                    break;
                }
                ptr = ptr->children[s[j]];
                if (ptr->terminal)
                {
                    dp[index] = min(dp[index], dp[j + 1]);
                }
            }
        }
        return dp[0];
    }
};

int main()
{
    cout << "Running Extra_Characters_In_String.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
